File: Medium/105. Construct Binary Tree from Preorder and Inorder Traversal/build_bst.cpp
#include <vector>
#include <unordered_map>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    int preorderIndex = 0; 

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n = inorder.size();
        unordered_map<int, int> inOrderMap;
        inOrderMap.reserve(n);

        for (int i = 0; i < n; i++) {
            inOrderMap.insert(make_pair(inorder[i], i));
        }

        return arrayToTree(0, n - 1, preorder, inorder, inOrderMap);

    }

    TreeNode* arrayToTree(int left, int right, vector<int>& preorder, vector<int>& inorder, unordered_map<int, int>& inOrderMap) {
        if (left > right) {
            return nullptr;
        }

        int rootVal = preorder[preorderIndex];
        TreeNode* root = &TreeNode(rootVal);
        preorderIndex++;

        root->left = arrayToTree(left, inOrderMap[rootVal] - 1, preorder, inorder);
        root->right = arrayToTree(inOrderMap[rootVal] + 1, right, preorder, inorder);

        return root;
    }
};

int main() {

    return 0;
}

File: Medium/105. Construct Binary Tree from Preorder and Inorder Traversal/build_bst.py
"""
Given two integer arrays preorder and inorder where preorder is the preorder traversal of a binary tree and inorder is the inorder traversal of the same tree, construct and return the binary tree.

1 <= preorder.length <= 3000
inorder.length == preorder.length
-3000 <= preorder[i], inorder[i] <= 3000
preorder and inorder consist of unique values.
Each value of inorder also appears in preorder.
preorder is guaranteed to be the preorder traversal of the tree.
inorder is guaranteed to be the inorder traversal of the tree.

"""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# preorder is root -> left -> right
# inorder is left -> root -> right 
# postorder is left -> right -> root

class Solution:
    def buildTree(self, preorder, inorder):
        preorderIndex = 0
        
        def arrayToTree(left, right):
            nonlocal preorderIndex
            
            # base case
            if left > right:
                return None

            # use the current element of preorder to define root of subtree
            rootVal = preorder[preorderIndex]
            root = TreeNode(val=rootVal)
            preorderIndex += 1

            # build the left and right subtrees. If no subtrees, we are at a leaf - so return
            root.left = arrayToTree(left, inOrderMap[rootVal] - 1)
            root.right = arrayToTree(inOrderMap[rootVal] + 1, right)
            return root
            
        # hashmap mapping value to index position so we can lookup the root in constant time
        inOrderMap = {k:v for (v,k) in enumerate(inorder)}

        return arrayToTree(0, len(inorder) - 1)

